src/app/main.rs

```diff
@@ -358,37 +335,62 @@ where
-                    let mut rt =
-                        current_thread::Runtime::new().expect("initialize admin thread runtime");
-
-                    rt.spawn(admin::serve_http(
-                        "admin",
-                        admin_listener,
-                        Admin::new(report, readiness, trace_level),
-                    ));
-
-                    if let Some((listener, tap_svc_name)) = control_listener {
-                        rt.spawn(tap_daemon.map_err(|_| ()));
-                        rt.spawn(serve_tap(listener, tap_svc_name, TapServer::new(tap_grpc)));
-                    }
-
-                    rt.spawn(logging::admin().bg("dns-resolver").future(dns_bg));
-
-                    if let Some(d) = identity_daemon {
-                        rt.spawn(
-                            logging::admin()
-                                .bg("identity")
-                                .future(d.map_err(|_| error!("identity task failed"))),
-                        );
-                    }
-
-                    let shutdown = admin_shutdown_signal.then(|_| Ok::<(), ()>(()));
-                    rt.block_on(shutdown).expect("admin");
+                    current_thread::Runtime::new()
+                        .expect("initialize admin thread runtime")
+                        .block_on(future::lazy(move || {
+                            trace!("spawning admin server");
+                            task::spawn(serve::serve(
+                                "admin",
+                                admin_listener,
+                                tls::AcceptTls::new(
+                                    get_original_dst.clone(),
+                                    local_identity.clone(),
+                                    Admin::new(report, readiness, trace_level).into_accept(),
+                                ),
+                                drain_rx.clone(),
+                            ));
+
+                            if let Some((listener, tap_svc_name)) = control_listener {
+                                trace!("spawning tap server");
+                                task::spawn(tap_daemon.map_err(|_| ()));
+                                task::spawn(serve::serve(
+                                    "tap",
+                                    listener,
+                                    tls::AcceptTls::new(
+                                        get_original_dst.clone(),
+                                        local_identity,
+                                        tap::AcceptPermittedClients::new(
+                                            std::iter::once(tap_svc_name),
+                                            tap_grpc,
+                                        ),
+                                    ),
+                                    drain_rx.clone(),
+                                ));
+                            } else {
+                                drop((tap_daemon, tap_grpc));
```

Comment:
This is good to see too. I don't think we were being explicit about this before.